pwndbg/aglib/stack.py

```diff
@@ -120,26 +127,50 @@ def _fetch_via_exploration() -> Dict[int, pwndbg.lib.memory.Page]:
     An alternative to this is dumping this functionality completely and this
     will be decided hopefully after a next release.
     """
-    stacks: Dict[int, pwndbg.lib.memory.Page] = {}
+    if auto_explore.value == "warn":
+        print(
+            M.warn(
+                "Warning: All methods to detect STACK have failed.\n"
+                "You can explore STACK using exploration, but it may be very slow.\n"
+                "To explicitly explore, use the command: `stack_explore`\n"
+                "Alternatively, enable it by default with: `set auto-explore-stack yes`"
+            )
+        )
+        return {}
+    elif auto_explore.value == "no":
+        return {}
 
-    for thread in pwndbg.dbg.selected_inferior().threads():
-        sp = thread.bottom_frame().sp()
+    thread_sp = []
+    for thread in reversed(pwndbg.dbg.selected_inferior().threads()):
+        with thread.bottom_frame() as frame:
+            sp = frame.sp()
 
         # Skip if sp is None or 0
         # (it might be 0 if we debug a qemu kernel)
         if not sp:
             continue
+        thread_sp.append((sp, thread.index()))
+
+    # Sort by the `sp` register (stack pointer), starting with the smallest value.
+    # This helps prevent scanning the same page multiple times.
+    thread_sp.sort(key=lambda t: t[0])
+
+    stacks: Dict[int, pwndbg.lib.memory.Page] = {}
+    for sp, thread_idx in thread_sp:
+        start = pwndbg.lib.memory.page_align(sp) - pwndbg.lib.memory.PAGE_SIZE
 
-        sp_low = sp & ~(0xFFF)
-        sp_low -= 0x1000
+        page_found = next((page for page in stacks.values() if start in page), None)
+        if page_found:
+            # Skip further exploration of stacks that have already been scanned.
+            stacks[thread_idx] = page_found
+            continue
 
-        start = sp_low
         stop = find_upper_stack_boundary(sp)
         page = pwndbg.lib.memory.Page(
-            start, stop - start, 6 if not is_executable() else 7, 0, f"[stack:{thread.index()}]"
+            start, stop - start, 6 if not is_executable() else 7, 0, "[stack]"
```

Comment:
Is this fine that we name this `[stack]` instead of `[stack:<tid>]`? What if there are multiple threads?

Reply:
In different places we have eg. `pwndbg.lib.memory.Page`:
<img width="459" alt="image" src="https://github.com/user-attachments/assets/75ced286-0b52-42cf-aca7-9faba8600108" />

Reply:
its hard to say whether this info `f"[stack:{thread.index()}]"` is useful?

Reply:
Maybe it should be `self.objfile.startswith("[stack")` eh..

Reply:
I think this is to color the main thread stack generally speaking, but its not great considering that multiple stacks can exist...

Reply:
done